### PR TITLE
Rewrite of app-scheduler to functional equivalent

### PR DESCRIPTION
**Pull request by scalvert**
This PR rewrites the app-scheduler to a functional equivalent. A new module is introduced that provides functional primitives which allow you to utilize a promise-based mechanism to approximate `afterFirstRoutePaint` and `afterContentPaint`, now named to `whenRoutePainted` and `whenRouteIdle` respectively.

This PR also removes the existing scheduler service, which is deprecated in favor of the promise based alternative.

TODO:
- [x] Write functional API
- [x] Add tests
- [x] Update README - [rendered](https://github.com/ember-app-scheduler/ember-app-scheduler/blob/remove-initializers/README.md).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -17,15 +17,15 @@ This addon provides a service to schedule work in different paint phases of rend
 The `afterFirstRoutePaint` queue can be used to schedule work after the route is first painted. This is useful for scenarios like rendering content outside viewport, rendering non critical content etc.
 
 ```javascript
-this.get('scheduler').on('afterFirstRoutePaint', () => {
```
Comment (runspired) [inline README.md]: YAAASS `'on'` made this seem evented.

File: addon/scheduler.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+function afterNextPaint() {
+  return new RSVP.Promise(resolve => {
+    if (USE_RAF) {
+      requestAnimationFrame(() => {
+        run.later(resolve, 0);
+      });
+    } else {
+      run.later(resolve, 0);
```
Comment (runspired) [inline addon/scheduler.js]: I would probably still try to defer this a bit, like `run.later(resolve, 16)`

File: addon/scheduler.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+let _didTransition;
+let _afterFirstRoutePaint;
+let _afterContentPaint;
+
+reset();
+
+export function beginTransition() {
+  _didTransition = RSVP.defer();
+  _afterFirstRoutePaint = _didTransition.promise.then(() => afterNextPaint());
+  _afterContentPaint = _afterFirstRoutePaint.then(() => afterNextPaint());
```
Comment (runspired) [inline addon/scheduler.js]: I suspect this queue is meaningless and not what we want and we can simply drop it from this re-imagined implementation.

Alternatively, we can fix it. It seems folks want the ability to schedule "do some work after the next paint" from an already active and rendered route. That is not what this queue accomplishes.

File: addon/scheduler.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+/**
+ * This promise, when resolved, approximates after content is painted.
+ */
+export function afterContentPaint() {
+  return _afterContentPaint;
+}
+
+function afterNextPaint() {
```
Comment (runspired) [inline addon/scheduler.js]: I like the function module-state pattern you've used, don't forget to cleanup the existing service that no longer needs this plumbing. I'm assuming that something in the consuming app will need to call `reset`?

File: addon/scheduler.js (hunk 3 above)
Comment (rwjblue) [reply addon/scheduler.js]: I want to change it to use requestIdleCallback if available

File: addon/scheduler.js (hunk 4 above)
Comment (rwjblue) [reply addon/scheduler.js]: @runspired - Ya, the idea is that the consuming app does `setupRouter(this)` in its `init` and `reset()` in its `destroy`.

Comment (runspired) [reply addon/scheduler.js]: @rwjblue we should add an assertion if you try to use the scheduler service without having done this.

File: addon/scheduler.js (hunk 3 above)
Comment (runspired) [reply addon/scheduler.js]: @rwjblue seems orthogonal to the question of whether this queue is even meaningful. At the moment it is implemented as `afterSecondRoutePaint` which is not what I think we want.

File: README.md (hunk 1 above)
Comment (scalvert) [reply README.md]: Yes, but the interface felt more natural.

File: addon/scheduler.js (hunk 2 above)
Comment (scalvert) [reply addon/scheduler.js]: I'm not sure that necessarily buys us anything, and though I see where you're going here, I'd rather have a concrete reason to do this. TL;DR - it feels like Thread.Sleep.

File: addon/scheduler.js (hunk 3 above)
Comment (scalvert) [reply addon/scheduler.js]: Yes, this is actually `do something after second paint`. What do you think we want here?

File: addon/scheduler.js (hunk 4 above)
Comment (scalvert) [reply addon/scheduler.js]: Already in our list!

File: addon/scheduler.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+reset();
+
+export const TRANSITION_INTERUPTED = 'TRANSITION INTERUPTED';
+
+export function beginTransition() {
+  _checkForPriorTransition();
+
+  _didTransition = _defer();
+  _whenRoutePainted = _didTransition.promise.then(() => _afterNextPaint());
+  _whenRouteIdle = _whenRoutePainted.then(() => _afterNextPaint());
```
Comment (runspired) [inline addon/scheduler.js]: I'm a fan. Still unconvinced it is a meaningful queue but at least now folks will have an easier time groking intent.

File: addon/scheduler.js (hunk 2 above)
Comment (runspired) [reply addon/scheduler.js]: We discussed this offline: TL;DR my concern was that `later(fn, 0)` might not preserve order of repeated calls due to the way backburner splices leading to odd timing issues. We confirmed this is not an issue, and `0` is just fine.

Comment (scalvert) [reply addon/scheduler.js]: 👍

**Pull request merged**